sandbox/lua/decoders/json.lua

```
@@ -0,0 +1,170 @@
+    | name:"extra.workspace-id" value_type:STRING value_string:"cN907xLngi"
+    | name:"time" value_type:STRING value_string:"2015-05-06T20:40:05.509926234Z"
+--]]
+
+require "cjson"
+local util = require("util")
+
+local msg_type     = read_config("type") or "json"
+local payload_keep = read_config("payload_keep")
+local map_fields   = read_config("map_fields")
+
+local field_map = {
+    Uuid       = read_config("Uuid"),
+    Type       = read_config("Type"),
+    Logger     = read_config("Logger"),
+    Hostname   = read_config("Hostname"),
+    Severity   = read_config("Severity"),
+    EnvVersion = read_config("EnvVersion"),
+    Pid        = read_config("Pid"),
+    Timestamp  = read_config("Timestamp")
+}
+
+local field_type_map = {
+    Uuid       = "string",
+    Type       = "string",
+    Logger     = "string",
+    Hostname   = "string",
+    Severity   = "number",
+    EnvVersion = "number",
+    Pid        = "number",
+    Timestamp  = "number"
+}
+
+local msg = {
+    Type     = msg_type,
+    Payload  = nil,
+    Fields   = nil,
+    Severity = nil
+}
+
+function process_message()
+    local ok, json = pcall(cjson.decode, read_message("Payload"))
+    if not ok then
+        return -1
+    end
+
+    -- keep payload, or not
+    if payload_keep then
+        msg.Payload = read_message("Payload")
+    end
+
+    -- map fields
+    if map_fields then
+        for F, f in pairs(field_map) do
+            if type(json[f]) == field_type_map[F] then
+                if F == "Timestamp" then
+                    msg[F] = json[f] * 1e9
+                else
+                    msg[F] = json[f]
+                end
+                json[f] = nil
+            end
+        end
+    end
+
+    -- flatten and assign remaining fields to heka fields
+    local flat = {}
+    util.table_to_fields(json, flat, nil)
+    msg.Fields = flat
+
+    if not pcall(inject_message, msg) then return -1 end
```

Another place we can return an error message.